## Practice Exercise 2

In this assignment, you will try to find some interesting insights into a few movies released between 1916 and 2016, using Python. You will have to download a movie dataset, write Python code to explore the data, gain insights into the movies, actors, directors, and collections, and submit the code.

#### Some tips before starting the assignment

1. Identify the task to be performed correctly, and only then proceed to write the required code. Don’t perform any incorrect analysis or look for information that isn’t required for the assignment.
2. In some cases, the variable names have already been assigned, and you just need to write code against them. In other cases, the names to be given are mentioned in the instructions. We strongly advise you to use the mentioned names only.
3. Always keep inspecting your data frame after you have performed a particular set of operations.
4. There are some checkpoints given in the IPython notebook provided. They're just useful pieces of information you can use to check if the result you have obtained after performing a particular task is correct or not.
5. Note that you will be asked to refer to documentation for solving some of the questions. That is done on purpose for you to learn new commands and also how to use the documentation.

In [2]:
# Import the numpy and pandas packages

import numpy as np
import pandas as pd

### Task 1: Reading and Inspection

**Subtask 1.1: Import and read**

Import and read the movie database. Store it in a variable called `movies`.

In [4]:
# Write your code for importing the csv file here
movies = pd.read_csv('Movie_dataset\Movies.csv')

movies.head(5)

color      director_name  num_critic_for_reviews  duration  \
0  Color      James Cameron                   723.0     178.0   
1  Color     Gore Verbinski                   302.0     169.0   
2  Color         Sam Mendes                   602.0     148.0   
3  Color  Christopher Nolan                   813.0     164.0   
4  Color     Andrew Stanton                   462.0     132.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   
3                  22000.0                 23000.0    Christian Bale   
4                    475.0                   530.0   Samantha Morton   

   actor_1_facebook_likes        gross                           genres  ...  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi  ...   
1                 40000.0  309404152.0         Action|Adventure|Fantasy  ...   
2                 11000.0  200074175.0        Action|Adventure|Thriller  ...   
3                 27000.0  448130642.0                  Action|Thriller  ...   
4                   640.0   73058679.0          Action|Adventure|Sci-Fi  ...   

  num_user_for_reviews language  country  content_rating       budget  \
0               3054.0  English      USA           PG-13  237000000.0   
1               1238.0  English      USA           PG-13  300000000.0   
2                994.0  English       UK           PG-13  245000000.0   
3               2701.0  English      USA           PG-13  250000000.0   
4                738.0  English      USA           PG-13  263700000.0   

   title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0      2009.0                  936.0        7.9          1.78   
1      2007.0                 5000.0        7.1          2.35   
2      2015.0                  393.0        6.8          2.35   
3      2012.0                23000.0        8.5          2.35   
4      2012.0                  632.0        6.6          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  
3               164000  
4                24000  

[5 rows x 28 columns]

**Subtask 1.2: Inspect the dataframe**

Inspect the dataframe's columns, shapes, variable types etc.

In [5]:
# Write your code for inspection here
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3853 entries, 0 to 3852
Data columns (total 28 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   color                      3851 non-null   object 
 1   director_name              3853 non-null   object 
 2   num_critic_for_reviews     3852 non-null   float64
 3   duration                   3852 non-null   float64
 4   director_facebook_likes    3853 non-null   float64
 5   actor_3_facebook_likes     3847 non-null   float64
 6   actor_2_name               3852 non-null   object 
 7   actor_1_facebook_likes     3853 non-null   float64
 8   gross                      3853 non-null   float64
 9   genres                     3853 non-null   object 
 10  actor_1_name               3853 non-null   object 
 11  movie_title                3853 non-null   object 
 12  num_voted_users            3853 non-null   int64  
 13  cast_total_facebook_likes  3853 non-null   int64

#### <font color='red'>Question 1: How many rows and columns are present in the dataframe? </font>
-  (3821, 26)
-  (3879, 28)
-  (3853, 28) -> Correct
-  (3866, 26)

#### <font color='red'>Question 2: How many columns have null values present in them? Try writing a code for this instead of counting them manually.</font>

-  <font color='red'>3</font>
-  <font color='red'>6</font>
-  <font color='red'>9</font>
-  <font color='red'>12</font> -> Correct

In [10]:
# Finding columns with Null values

print((movies.isnull().sum()>0).sum())

12


### Task 2: Cleaning the Data

**Subtask 2.1: Drop unecessary columns**

For this assignment, you will mostly be analyzing the movies with respect to the ratings, gross collection, popularity of movies, etc.<br>So many of the columns in this dataframe are not required.<br>So it is advised to drop the following columns.
-  color
-  director_facebook_likes
-  actor_1_facebook_likes
-  actor_2_facebook_likes
-  actor_3_facebook_likes
-  actor_2_name
-  cast_total_facebook_likes
-  actor_3_name
-  duration
-  facenumber_in_poster
-  content_rating
-  country
-  movie_imdb_link
-  aspect_ratio
-  plot_keywords

In [12]:
# Check the 'drop' function in the Pandas library - dataframe.drop(list_of_unnecessary_columns, axis = )
# Write your code for dropping the columns here. It is advised to keep inspecting the dataframe after each set of operations

movies.drop(['color', 'director_facebook_likes', 'actor_1_facebook_likes', 'actor_2_facebook_likes',
            'actor_3_facebook_likes', 'actor_2_name', 'cast_total_facebook_likes', 'actor_3_name',
             'duration', 'facenumber_in_poster', 'content_rating', 'country', 'movie_imdb_link',
              'aspect_ratio', 'plot_keywords'], axis=1, inplace=True)

movies

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes  
0                    33000  
1                        0  
2                    85000  
3                   164000  
4                    24000  
...                    ...  
3848                 19000  
3849                    74  
3850                     0  
3851                   413  
3852                   456  

[3853 rows x 13 columns]

**Subtask 2.2: Inspect Null values**

As you have seen above, there are null values in multiple columns of the dataframe 'movies'. Find out the percentage of null values in each column of the dataframe 'movies'. 

In [15]:
# Write you code here

round(100*(movies.isnull().sum()/len(movies.index)), 2)

director_name             0.00
num_critic_for_reviews    0.03
gross                     0.00
genres                    0.00
actor_1_name              0.00
movie_title               0.00
num_voted_users           0.00
num_user_for_reviews      0.00
language                  0.10
budget                    0.00
title_year                0.00
imdb_score                0.00
movie_facebook_likes      0.00
dtype: float64

In [16]:
movies.isnull().sum()

director_name             0
num_critic_for_reviews    1
gross                     0
genres                    0
actor_1_name              0
movie_title               0
num_voted_users           0
num_user_for_reviews      0
language                  4
budget                    0
title_year                0
imdb_score                0
movie_facebook_likes      0
dtype: int64

#### <font color='red'>Question 4: Which column has the highest percentage of null values? </font>
-  <font color='red'>language</font> -> Correct
-  <font color='red'>genres</font>
-  <font color='red'>num_critic_for_reviews</font>
-  <font color='red'>imdb_score</font>

**Subtask 2.3: Fill NaN values**

You might notice that the `language` column has some NaN values. Here, on inspection, you will see that it is safe to replace all the missing values with `'English'`.

In [17]:
# Write your code for filling the NaN values in the 'language' column here

movies['language'].fillna('English', inplace=True)

In [18]:
movies['language'].value_counts()

language
English       3675
French          37
Spanish         26
Mandarin        14
German          13
Japanese        12
Hindi           10
Cantonese        8
Italian          7
Portuguese       5
Korean           5
Norwegian        4
Thai             3
Persian          3
Danish           3
Dutch            3
Dari             2
Indonesian       2
Hebrew           2
Aboriginal       2
Arabic           1
Russian          1
Vietnamese       1
Dzongkha         1
Romanian         1
Zulu             1
Bosnian          1
Czech            1
Icelandic        1
Hungarian        1
Mongolian        1
Aramaic          1
Telugu           1
Kazakh           1
Maya             1
Filipino         1
Swedish          1
Name: count, dtype: int64

#### <font color='red'>Question 5: What is the count of movies made in English language after replacing the NaN values with English? </font>
-  <font color='red'>3670</font>
-  <font color='red'>3674</font>
-  <font color='red'>3668</font>
-  <font color='red'>3672</font>

### Task 3: Data Analysis

**Subtask 3.1: Change the unit of columns**

Convert the unit of the `budget` and `gross` columns from `$` to `million $`.

In [ ]:
# Write your code for unit conversion here


**Subtask 3.2: Find the movies with highest profit**

   1. Create a new column called `profit` which contains the difference of the two columns: `gross` and `budget`.
   2. Sort the dataframe using the `profit` column as reference. (Find which command can be used here to sort entries from the documentation)
   3. Extract the top ten profiting movies in descending order and store them in a new dataframe - `top10`

In [19]:
# Write your code for creating the profit column here
movies['net_profit'] = movies['gross'] - movies['budget']

movies

director_name  num_critic_for_reviews        gross  \
0         James Cameron                   723.0  760505847.0   
1        Gore Verbinski                   302.0  309404152.0   
2            Sam Mendes                   602.0  200074175.0   
3     Christopher Nolan                   813.0  448130642.0   
4        Andrew Stanton                   462.0   73058679.0   
...                 ...                     ...          ...   
3848      Shane Carruth                   143.0     424760.0   
3849   Neill Dela Llana                    35.0      70071.0   
3850   Robert Rodriguez                    56.0    2040920.0   
3851       Edward Burns                    14.0       4584.0   
3852           Jon Gunn                    43.0      85222.0   

                                   genres     actor_1_name  \
0         Action|Adventure|Fantasy|Sci-Fi      CCH Pounder   
1                Action|Adventure|Fantasy      Johnny Depp   
2               Action|Adventure|Thriller  Christoph Waltz   
3                         Action|Thriller        Tom Hardy   
4                 Action|Adventure|Sci-Fi     Daryl Sabara   
...                                   ...              ...   
3848                Drama|Sci-Fi|Thriller    Shane Carruth   
3849                             Thriller      Ian Gamazon   
3850  Action|Crime|Drama|Romance|Thriller  Carlos Gallardo   
3851                         Comedy|Drama      Kerry Bishé   
3852                          Documentary      John August   

                                    movie_title  num_voted_users  \
0                                       Avatar            886204   
1     Pirates of the Caribbean: At World's End            471220   
2                                      Spectre            275868   
3                        The Dark Knight Rises           1144337   
4                                  John Carter            212204   
...                                         ...              ...   
3848                                    Primer             72639   
3849                                    Cavite               589   
3850                               El Mariachi             52055   
3851                                 Newlyweds              1338   
3852                         My Date with Drew              4285   

      num_user_for_reviews language       budget  title_year  imdb_score  \
0                   3054.0  English  237000000.0      2009.0         7.9   
1                   1238.0  English  300000000.0      2007.0         7.1   
2                    994.0  English  245000000.0      2015.0         6.8   
3                   2701.0  English  250000000.0      2012.0         8.5   
4                    738.0  English  263700000.0      2012.0         6.6   
...                    ...      ...          ...         ...         ...   
3848                 371.0  English       7000.0      2004.0         7.0   
3849                  35.0  English       7000.0      2005.0         6.3   
3850                 130.0  Spanish       7000.0      1992.0         6.9   
3851                  14.0  English       9000.0      2011.0         6.4   
3852                  84.0  English       1100.0      2004.0         6.6   

      movie_facebook_likes   net_profit  
0                    33000  523505847.0  
1                        0    9404152.0  
2                    85000  -44925825.0  
3                   164000  198130642.0  
4                    24000 -190641321.0  
...                    ...          ...  
3848                 19000     417760.0  
3849                    74      63071.0  
3850                     0    2033920.0  
3851                   413      -4416.0  
3852                   456      84122.0  

[3853 rows x 14 columns]

In [20]:
# Write your code for sorting the dataframe here
movies.sort_values('net_profit', ascending=False)

director_name  num_critic_for_reviews        gross  \
0        James Cameron                   723.0  760505847.0   
28     Colin Trevorrow                   644.0  652177271.0   
25       James Cameron                   315.0  658672302.0   
2704      George Lucas                   282.0  460935665.0   
2748  Steven Spielberg                   215.0  434949459.0   
...                ...                     ...          ...   
2147   Katsuhiro Ôtomo                   105.0     410388.0   
2136    Hayao Miyazaki                   174.0    2298191.0   
2693      Lajos Koltai                    73.0     195888.0   
3282    Chan-wook Park                   202.0     211667.0   
2678      Joon-ho Bong                   363.0    2201412.0   

                                                 genres         actor_1_name  \
0                       Action|Adventure|Fantasy|Sci-Fi          CCH Pounder   
28                     Action|Adventure|Sci-Fi|Thriller  Bryce Dallas Howard   
25                                        Drama|Romance    Leonardo DiCaprio   
2704                    Action|Adventure|Fantasy|Sci-Fi        Harrison Ford   
2748                                      Family|Sci-Fi         Henry Thomas   
...                                                 ...                  ...   
2147  Action|Adventure|Animation|Family|Sci-Fi|Thriller     William Hootkins   
2136                        Adventure|Animation|Fantasy        Minnie Driver   
2693                                  Drama|Romance|War         Marcell Nagy   
3282                                        Crime|Drama         Min-sik Choi   
2678                         Comedy|Drama|Horror|Sci-Fi            Doona Bae   

                              movie_title  num_voted_users  \
0                                 Avatar            886204   
28                        Jurassic World            418214   
25                               Titanic            793059   
2704  Star Wars: Episode IV - A New Hope            911097   
2748          E.T. the Extra-Terrestrial            281842   
...                                   ...              ...   
2147                            Steamboy             13727   
2136                   Princess Mononoke            221552   
2693                            Fateless              5603   
3282                      Lady Vengeance             53508   
2678                            The Host             68883   

      num_user_for_reviews   language        budget  title_year  imdb_score  \
0                   3054.0    English  2.370000e+08      2009.0         7.9   
28                  1290.0    English  1.500000e+08      2015.0         7.0   
25                  2528.0    English  2.000000e+08      1997.0         7.7   
2704                1470.0    English  1.100000e+07      1977.0         8.7   
2748                 515.0    English  1.050000e+07      1982.0         7.9   
...                    ...        ...           ...         ...         ...   
2147                  79.0   Japanese  2.127520e+09      2004.0         6.9   
2136                 570.0   Japanese  2.400000e+09      1997.0         8.4   
2693                  45.0  Hungarian  2.500000e+09      2005.0         7.1   
3282                 131.0     Korean  4.200000e+09      2005.0         7.7   
2678                 279.0     Korean  1.221550e+10      2006.0         7.0   

      movie_facebook_likes    net_profit  
0                    33000  5.235058e+08  
28                  150000  5.021773e+08  
25                   26000  4.586723e+08  
2704                 33000  4.499357e+08  
2748                 34000  4.244495e+08  
...                    ...           ...  
2147                   973 -2.127110e+09  
2136                 11000 -2.397702e+09  
2693                   607 -2.499804e+09  
3282                  4000 -4.199788e+09  
2678                  7000 -1.221330e+10  

[3853 rows x 14 columns]

In [ ]:
top10 = # Write your code to get the top 10 profiting movies here

**Checkpoint:** You might spot two movies directed by `James Cameron` in the list.

#### <font color='red'>Question 6: Which movie is ranked 5th from the top in the list obtained? </font>
-  <font color='red'>E.T. the Extra-Terrestrial</font> -> Correct
-  <font color='red'>The Avengers</font>
-  <font color='red'>The Dark Knight</font>
-  <font color='red'>Titanic</font>

**Subtask 3.3: Find IMDb Top 250**

Create a new dataframe `IMDb_Top_250` and store the top 250 movies with the highest IMDb Rating (corresponding to the column: `imdb_score`).<br>Also make sure that for all of these movies, the `num_voted_users` is greater than 25,000. 

Also add a `Rank` column containing the values 1 to 250 indicating the ranks of the corresponding films.

In [27]:
# Write your code for extracting the top 250 movies as per the IMDb score here. Make sure that you store it in a new dataframe 
# and name that dataframe as 'IMDb_Top_250'

IMDb_Top_250 = movies[movies['num_voted_users'] > 25000].sort_values('imdb_score', ascending=False).head(250)

director_name  num_critic_for_reviews        gross  \
1795        Frank Darabont                   199.0   28341469.0   
3016  Francis Ford Coppola                   208.0  134821952.0   
64       Christopher Nolan                   645.0  533316061.0   
2543  Francis Ford Coppola                   149.0   57300000.0   
325          Peter Jackson                   328.0  377019252.0   
...                    ...                     ...          ...   
67             Jon Favreau                   486.0  318298180.0   
73              Doug Liman                   585.0  100189501.0   
3052           Yash Chopra                    29.0    2921738.0   
95           Peter Jackson                   645.0  303001229.0   
3170  Christophe Barratier                   112.0    3629758.0   

                              genres           actor_1_name  \
1795                     Crime|Drama         Morgan Freeman   
3016                     Crime|Drama              Al Pacino   
64       Action|Crime|Drama|Thriller         Christian Bale   
2543                     Crime|Drama         Robert De Niro   
325   Action|Adventure|Drama|Fantasy          Orlando Bloom   
...                              ...                    ...   
67           Action|Adventure|Sci-Fi      Robert Downey Jr.   
73           Action|Adventure|Sci-Fi             Tom Cruise   
3052           Drama|Musical|Romance         Shah Rukh Khan   
95                 Adventure|Fantasy           Aidan Turner   
3170                     Drama|Music  Jean-Baptiste Maunier   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
64                                  The Dark Knight           1676169   
2543                         The Godfather: Part II            790926   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
67                                         Iron Man            696338   
73                                 Edge of Tomorrow            431620   
3052                                     Veer-Zaara             34449   
95                The Hobbit: An Unexpected Journey            637246   
3170                                     The Chorus             44151   

      num_user_for_reviews language       budget  title_year  imdb_score  \
1795                4144.0  English   25000000.0      1994.0         9.3   
3016                2238.0  English    6000000.0      1972.0         9.2   
64                  4667.0  English  185000000.0      2008.0         9.0   
2543                 650.0  English   13000000.0      1974.0         9.0   
325                 3189.0  English   94000000.0      2003.0         8.9   
...                    ...      ...          ...         ...         ...   
67                  1055.0  English  140000000.0      2008.0         7.9   
73                   741.0  English  178000000.0      2014.0         7.9   
3052                 119.0    Hindi    7000000.0      2004.0         7.9   
95                  1367.0  English  180000000.0      2012.0         7.9   
3170                 110.0   French    5500000.0      2004.0         7.9   

      movie_facebook_likes   net_profit  
1795                108000    3341469.0  
3016                 43000  128821952.0  
64                   37000  348316061.0  
2543                 14000   44300000.0  
325                  16000  283019252.0  
...                    ...          ...  
67                   10000  178298180.0  
73                   77000  -77810499.0  
3052                  2000   -4078262.0  
95                  166000  123001229.0  
3170                     0   -1870242.0  

[250 rows x 14 columns]

In [29]:
# Addign rank column

IMDb_Top_250['Rank'] = range(1, len(IMDb_Top_250) + 1)

IMDb_Top_250

director_name  num_critic_for_reviews        gross  \
1795        Frank Darabont                   199.0   28341469.0   
3016  Francis Ford Coppola                   208.0  134821952.0   
64       Christopher Nolan                   645.0  533316061.0   
2543  Francis Ford Coppola                   149.0   57300000.0   
325          Peter Jackson                   328.0  377019252.0   
...                    ...                     ...          ...   
67             Jon Favreau                   486.0  318298180.0   
73              Doug Liman                   585.0  100189501.0   
3052           Yash Chopra                    29.0    2921738.0   
95           Peter Jackson                   645.0  303001229.0   
3170  Christophe Barratier                   112.0    3629758.0   

                              genres           actor_1_name  \
1795                     Crime|Drama         Morgan Freeman   
3016                     Crime|Drama              Al Pacino   
64       Action|Crime|Drama|Thriller         Christian Bale   
2543                     Crime|Drama         Robert De Niro   
325   Action|Adventure|Drama|Fantasy          Orlando Bloom   
...                              ...                    ...   
67           Action|Adventure|Sci-Fi      Robert Downey Jr.   
73           Action|Adventure|Sci-Fi             Tom Cruise   
3052           Drama|Musical|Romance         Shah Rukh Khan   
95                 Adventure|Fantasy           Aidan Turner   
3170                     Drama|Music  Jean-Baptiste Maunier   

                                         movie_title  num_voted_users  \
1795                       The Shawshank Redemption           1689764   
3016                                  The Godfather           1155770   
64                                  The Dark Knight           1676169   
2543                         The Godfather: Part II            790926   
325   The Lord of the Rings: The Return of the King           1215718   
...                                              ...              ...   
67                                         Iron Man            696338   
73                                 Edge of Tomorrow            431620   
3052                                     Veer-Zaara             34449   
95                The Hobbit: An Unexpected Journey            637246   
3170                                     The Chorus             44151   

      num_user_for_reviews language       budget  title_year  imdb_score  \
1795                4144.0  English   25000000.0      1994.0         9.3   
3016                2238.0  English    6000000.0      1972.0         9.2   
64                  4667.0  English  185000000.0      2008.0         9.0   
2543                 650.0  English   13000000.0      1974.0         9.0   
325                 3189.0  English   94000000.0      2003.0         8.9   
...                    ...      ...          ...         ...         ...   
67                  1055.0  English  140000000.0      2008.0         7.9   
73                   741.0  English  178000000.0      2014.0         7.9   
3052                 119.0    Hindi    7000000.0      2004.0         7.9   
95                  1367.0  English  180000000.0      2012.0         7.9   
3170                 110.0   French    5500000.0      2004.0         7.9   

      movie_facebook_likes   net_profit  Rank  
1795                108000    3341469.0     1  
3016                 43000  128821952.0     2  
64                   37000  348316061.0     3  
2543                 14000   44300000.0     4  
325                  16000  283019252.0     5  
...                    ...          ...   ...  
67                   10000  178298180.0   246  
73                   77000  -77810499.0   247  
3052                  2000   -4078262.0   248  
95                  166000  123001229.0   249  
3170                     0   -1870242.0   250  

[250 rows x 15 columns]

#### <font color='red'>Question 7: Suppose movies are divided into 5 buckets based on the IMDb ratings: </font>
-  <font color='red'>7.5 to 8</font>
-  <font color='red'>8 to 8.5</font>
-  <font color='red'>8.5 to 9</font>
-  <font color='red'>9 to 9.5</font>
-  <font color='red'>9.5 to 10</font>

<font color = 'red'> Which bucket holds the maximum number of movies from *IMDb_Top_250*? </font>

In Pandas, the ``pd.cut()`` function is a powerful tool for binning or categorizing continuous numerical data into discrete intervals. It's particularly useful when you want to analyze data based on ranges or groups.

Key Parameters:

x: The input array to be binned.<br>
bins: 
>- Integer: Defines the number of equal-width bins.<br>
>- Array-like: Defines the bin edges.<br>

labels: Optional labels for the bins.<br>
right: Boolean indicating whether the intervals include the rightmost edge or not.<br>
include_lowest: Boolean indicating whether the leftmost interval should include its lower bound.<br>
retbins: Boolean indicating whether to return the bins along with the categorized Series.

In [32]:
# Create bins for rating groups
bins = [7.5, 8, 8.5, 9, 9.5, 10]
labels = ['7.5-8', '8-8.5', '8.5-9', '9-9.5', '9.5-10']

# Categorize movies based on rating
IMDb_Top_250['Rating_Group'] = pd.cut(IMDb_Top_250['imdb_score'], bins=bins, labels=labels, right=False)

# Group by rating group and count movies
grouped_df = IMDb_Top_250.groupby('Rating_Group').size()

# Find the group with the highest number of movies
highest_group = grouped_df.idxmax()
highest_count = grouped_df.max()

In [33]:
print(highest_group)
print(highest_count)

8-8.5
159


**Subtask 3.4: Find the critic-favorite and audience-favorite actors**

   1. Create three new dataframes namely, `Meryl_Streep`, `Leo_Caprio`, and `Brad_Pitt` which contain the movies in which the actors: 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' are the lead actors. Use only the `actor_1_name` column for extraction. Also, make sure that you use the names 'Meryl Streep', 'Leonardo DiCaprio', and 'Brad Pitt' for the said extraction.
   2. Append the rows of all these dataframes and store them in a new dataframe named `Combined`.
   3. Group the combined dataframe using the `actor_1_name` column.
   4. Find the mean of the `num_critic_for_reviews` and `num_user_for_review` and identify the actors which have the highest mean.

In [35]:
# Write your code for creating three new dataframes here
Meryl_Streep = movies[movies['actor_1_name'] == 'Meryl Streep']
Meryl_Streep

director_name  num_critic_for_reviews        gross  \
392     Nancy Meyers                   187.0  112703470.0   
1038   Curtis Hanson                    42.0   46815748.0   
1132     Nora Ephron                   252.0   94125426.0   
1322   David Frankel                   208.0  124732962.0   
1390  Robert Redford                   227.0   14998070.0   
1471  Sydney Pollack                    66.0   87100000.0   
1514   David Frankel                   234.0   63536011.0   
1563   Carl Franklin                    64.0   23209440.0   
1784  Stephen Daldry                   174.0   41597830.0   
2500  Phyllida Lloyd                   331.0   29959436.0   
2793   Robert Altman                   211.0   20338609.0   

                               genres  actor_1_name  \
392              Comedy|Drama|Romance  Meryl Streep   
1038  Action|Adventure|Crime|Thriller  Meryl Streep   
1132          Biography|Drama|Romance  Meryl Streep   
1322             Comedy|Drama|Romance  Meryl Streep   
1390               Drama|Thriller|War  Meryl Streep   
1471          Biography|Drama|Romance  Meryl Streep   
1514             Comedy|Drama|Romance  Meryl Streep   
1563                            Drama  Meryl Streep   
1784                    Drama|Romance  Meryl Streep   
2500          Biography|Drama|History  Meryl Streep   
2793               Comedy|Drama|Music  Meryl Streep   

                    movie_title  num_voted_users  num_user_for_reviews  \
392           It's Complicated             69860                 214.0   
1038            The River Wild             32544                  69.0   
1132             Julie & Julia             79264                 277.0   
1322     The Devil Wears Prada            286178                 631.0   
1390           Lions for Lambs             41170                 298.0   
1471             Out of Africa             52339                 200.0   
1514              Hope Springs             34258                 178.0   
1563            One True Thing              9283                 112.0   
1784                 The Hours            102123                 660.0   
2500             The Iron Lady             82327                 350.0   
2793  A Prairie Home Companion             19655                 280.0   

     language      budget  title_year  imdb_score  movie_facebook_likes  \
392   English  85000000.0      2009.0         6.6                     0   
1038  English  45000000.0      1994.0         6.3                     0   
1132  English  40000000.0      2009.0         7.0                 13000   
1322  English  35000000.0      2006.0         6.8                     0   
1390  English  35000000.0      2007.0         6.2                     0   
1471  English  31000000.0      1985.0         7.2                     0   
1514  English  30000000.0      2012.0         6.3                     0   
1563  English  30000000.0      1998.0         7.0                   592   
1784  English  25000000.0      2002.0         7.6                     0   
2500  English  13000000.0      2011.0         6.4                 18000   
2793  English  10000000.0      2006.0         6.8                   683   

      net_profit Rating_Group  
392   27703470.0          NaN  
1038   1815748.0          NaN  
1132  54125426.0          NaN  
1322  89732962.0          NaN  
1390 -20001930.0          NaN  
1471  56100000.0          NaN  
1514  33536011.0          NaN  
1563  -6790560.0          NaN  
1784  16597830.0        7.5-8  
2500  16959436.0          NaN  
2793  10338609.0          NaN

In [41]:
Leo_Caprio = movies[movies['actor_1_name'] == 'Leonardo DiCaprio']
Leo_Caprio

director_name  num_critic_for_reviews        gross  \
25            James Cameron                   315.0  658672302.0   
49             Baz Luhrmann                   490.0  144812796.0   
94        Christopher Nolan                   642.0  292568851.0   
173   Alejandro G. Iñárritu                   556.0  183635922.0   
246         Martin Scorsese                   267.0  102608827.0   
283       Quentin Tarantino                   765.0  162804648.0   
293            Edward Zwick                   166.0   57366262.0   
294         Martin Scorsese                   606.0  116866727.0   
312         Martin Scorsese                   233.0   77679638.0   
347         Martin Scorsese                   352.0  132373442.0   
431         Martin Scorsese                   490.0  127968405.0   
608            Ridley Scott                   238.0   39380442.0   
859        Steven Spielberg                   194.0  164435221.0   
934             Danny Boyle                   118.0   39778599.0   
1046             Sam Mendes                   323.0   22877808.0   
1336        Randall Wallace                    83.0   56876365.0   
1364         Clint Eastwood                   392.0   37304950.0   
1457              Sam Raimi                    63.0   18636537.0   
1909             Jerry Zaks                    45.0   12782508.0   
2484           Baz Luhrmann                   106.0   46338728.0   
3025           Baz Luhrmann                   490.0  144812796.0   

                                genres       actor_1_name  \
25                       Drama|Romance  Leonardo DiCaprio   
49                       Drama|Romance  Leonardo DiCaprio   
94    Action|Adventure|Sci-Fi|Thriller  Leonardo DiCaprio   
173   Adventure|Drama|Thriller|Western  Leonardo DiCaprio   
246                    Biography|Drama  Leonardo DiCaprio   
283                      Drama|Western  Leonardo DiCaprio   
293           Adventure|Drama|Thriller  Leonardo DiCaprio   
294       Biography|Comedy|Crime|Drama  Leonardo DiCaprio   
312                        Crime|Drama  Leonardo DiCaprio   
347               Crime|Drama|Thriller  Leonardo DiCaprio   
431                   Mystery|Thriller  Leonardo DiCaprio   
608              Action|Drama|Thriller  Leonardo DiCaprio   
859              Biography|Crime|Drama  Leonardo DiCaprio   
934           Adventure|Drama|Thriller  Leonardo DiCaprio   
1046                     Drama|Romance  Leonardo DiCaprio   
1336                  Action|Adventure  Leonardo DiCaprio   
1364             Biography|Crime|Drama  Leonardo DiCaprio   
1457           Action|Thriller|Western  Leonardo DiCaprio   
1909                             Drama  Leonardo DiCaprio   
2484                     Drama|Romance  Leonardo DiCaprio   
3025                     Drama|Romance  Leonardo DiCaprio   

                    movie_title  num_voted_users  num_user_for_reviews  \
25                     Titanic            793059                2528.0   
49            The Great Gatsby            362912                 753.0   
94                   Inception           1468200                2803.0   
173               The Revenant            406020                1188.0   
246                The Aviator            264318                 799.0   
283           Django Unchained            955174                1193.0   
293              Blood Diamond            400292                 657.0   
294    The Wolf of Wall Street            780588                1138.0   
312          Gangs of New York            314033                1166.0   
347               The Departed            873649                2054.0   
431             Shutter Island            786092                 964.0   
608               Body of Lies            174248                 263.0   
859        Catch Me If You Can            525801                 667.0   
934                  The Beach            176169                 548.0   
1046        Revolutionary Road            152591                

In [39]:
Brad_Pitt = movies[movies['actor_1_name'] == 'Brad Pitt']
Brad_Pitt

director_name  num_critic_for_reviews        gross  \
97            David Fincher                   362.0  127490802.0   
142       Wolfgang Petersen                   220.0  133228348.0   
243       Steven Soderbergh                   198.0  125531634.0   
244              Doug Liman                   233.0  186336103.0   
367              Tony Scott                   142.0      26871.0   
383       Steven Soderbergh                   186.0  183405771.0   
448              David Ayer                   406.0   85707116.0   
579     Jean-Jacques Annaud                    76.0   37901509.0   
646           David Fincher                   315.0   37023395.0   
749         Patrick Gilmore                    98.0   26288320.0   
887             Neil Jordan                   120.0  105264608.0   
1397        Terrence Malick                   584.0   13303319.0   
1608         Andrew Dominik                   273.0    3904982.0   
2030  Alejandro G. Iñárritu                   285.0   34300771.0   
2146    Angelina Jolie Pitt                   131.0     531009.0   
2428         Andrew Dominik                   414.0   14938570.0   
2602             Tony Scott                   122.0   12281500.0   

                                                 genres actor_1_name  \
97                                Drama|Fantasy|Romance    Brad Pitt   
142                                           Adventure    Brad Pitt   
243                                      Crime|Thriller    Brad Pitt   
244                Action|Comedy|Crime|Romance|Thriller    Brad Pitt   
367                               Action|Crime|Thriller    Brad Pitt   
383                                      Crime|Thriller    Brad Pitt   
448                                    Action|Drama|War    Brad Pitt   
579               Adventure|Biography|Drama|History|War    Brad Pitt   
646                                               Drama    Brad Pitt   
749   Adventure|Animation|Comedy|Drama|Family|Fantas...    Brad Pitt   
887                                Drama|Fantasy|Horror    Brad Pitt   
1397                                      Drama|Fantasy    Brad Pitt   
1608              Biography|Crime|Drama|History|Western    Brad Pitt   
2030                                              Drama    Brad Pitt   
2146                                      Drama|Romance    Brad Pitt   
2428                                     Crime|Thriller    Brad Pitt   
2602                Action|Crime|Drama|Romance|Thriller    Brad Pitt   

                                            movie_title  num_voted_users  \
97                 The Curious Case of Benjamin Button            459346   
142                                               Troy            381672   
243                                     Ocean's Twelve            284852   
244                                   Mr. & Mrs. Smith            348861   
367                                           Spy Game            121259   
383                                     Ocean's Eleven            402645   
448                                               Fury            303185   
579                               Seven Years in Tibet             96385   
646                                         Fight Club           1347461   
749                   Sinbad: Legend of the Seven Seas             36144   
887   Interview with the Vampire: The Vampire Chroni...           239752   
1397                                  The Tree of Life            136367   
1608  The Assassination of Jesse James by the Coward...           136104   
2030                                             Babel            243799   
2146                                        By the Sea              7976   
2428                               Killing Them Softly            111625   
2602                                      True Romance            163492   

      num_user_for_reviews language       budget  title_year  imdb_score  \
97                   822.0  English  150000000.

In [43]:
# Write your code for combining the three dataframes here
Combined_df = pd.concat([Meryl_Streep, Brad_Pitt, Leo_Caprio], axis=0)

Combined_df.head(5)

director_name  num_critic_for_reviews        gross  \
392     Nancy Meyers                   187.0  112703470.0   
1038   Curtis Hanson                    42.0   46815748.0   
1132     Nora Ephron                   252.0   94125426.0   
1322   David Frankel                   208.0  124732962.0   
1390  Robert Redford                   227.0   14998070.0   

                               genres  actor_1_name             movie_title  \
392              Comedy|Drama|Romance  Meryl Streep       It's Complicated    
1038  Action|Adventure|Crime|Thriller  Meryl Streep         The River Wild    
1132          Biography|Drama|Romance  Meryl Streep          Julie & Julia    
1322             Comedy|Drama|Romance  Meryl Streep  The Devil Wears Prada    
1390               Drama|Thriller|War  Meryl Streep        Lions for Lambs    

      num_voted_users  num_user_for_reviews language      budget  title_year  \
392             69860                 214.0  English  85000000.0      2009.0   
1038            32544                  69.0  English  45000000.0      1994.0   
1132            79264                 277.0  English  40000000.0      2009.0   
1322           286178                 631.0  English  35000000.0      2006.0   
1390            41170                 298.0  English  35000000.0      2007.0   

      imdb_score  movie_facebook_likes  net_profit Rating_Group  
392          6.6                     0  27703470.0          NaN  
1038         6.3                     0   1815748.0          NaN  
1132         7.0                 13000  54125426.0          NaN  
1322         6.8                     0  89732962.0          NaN  
1390         6.2                     0 -20001930.0          NaN

In [44]:
# Write your code for grouping the combined dataframe here
Combined_df.groupby('actor_1_name').mean(['num_critic_for_reviews', 'num_user_for_reviews'])

num_critic_for_reviews         gross  num_voted_users  \
actor_1_name                                                               
Brad Pitt                      245.000000  6.632145e+07    283583.823529   
Leonardo DiCaprio              330.190476  1.257420e+08    441958.857143   
Meryl Streep                   181.454545  5.991973e+07     73545.545455   

                   num_user_for_reviews        budget   title_year  \
actor_1_name                                                         
Brad Pitt                    742.352941  6.929412e+07  2004.352941   
Leonardo DiCaprio            914.476190  8.245238e+07  2005.476190   
Meryl Streep                 297.181818  3.445455e+07  2003.545455   

                   imdb_score  movie_facebook_likes    net_profit  
actor_1_name                                                       
Brad Pitt            7.152941          14051.764706 -2.972669e+06  
Leonardo DiCaprio    7.495238          52190.476190  4.328961e+07  
Meryl Streep         6.745455           2934.090909  2.546518e+07

In [ ]:
# Write the code for finding the mean of critic reviews and audience reviews here


#### <font color='red'>Question 8: Which actor is highest rated among the three actors according to the user reviews? </font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>

#### <font color='red'>Question 9: Which actor is highest rated among the three actors according to the critics?</font>
-  <font color='red'>Meryl Streep</font>
-  <font color='red'>Leonardo DiCaprio</font>
-  <font color='red'>Brad Pitt</font>